packages/labs/src/multi-select/multiSelect.tsx

```diff
@@ -0,0 +1,299 @@
+/*
+ * Copyright 2017 Palantir Technologies, Inc. All rights reserved.
+ * Licensed under the BSD-3 License as modified (the “License”); you may obtain a copy
+ * of the license at https://github.com/palantir/blueprint/blob/master/LICENSE
+ * and https://github.com/palantir/blueprint/blob/master/PATENTS
+ */
+
+import * as classNames from "classnames";
+import * as PureRender from "pure-render-decorator";
+import * as React from "react";
+
+import {
+    AbstractComponent,
+    Classes,
+    HTMLInputProps,
+    IPopoverProps,
+    Keys,
+    Popover,
+    Position,
+    Utils,
+} from "@blueprintjs/core";
+import { IListItemsProps, IQueryListRendererProps, QueryList } from "../query-list/queryList";
+import { ISelectItemRendererProps } from "../select/select";
+import { ITagInputProps, TagInput } from "../tag-input/tagInput";
+
+export interface IMultiSelectProps<T> extends IListItemsProps<T> {
+    /**
+     * Custom renderer for an item in the dropdown list. Receives a boolean indicating whether
+     * this item is active (selected by keyboard arrows) and an `onClick` event handler that
+     * should be attached to the returned element.
+     */
+    itemRenderer: (itemProps: IMultiSelectItemRendererProps<T>) => JSX.Element;
+
+    /** React child to render when filtering items returns zero results. */
+    noResults?: string | JSX.Element;
+
+    /**
+     * Whether the popover opens on key down or when `TagInput` is focused.
+     * @default false
+     */
+    openOnKeyDown?: boolean;
+
+    /** Props to spread to `Popover`. Note that `content` cannot be changed. */
+    popoverProps?: Partial<IPopoverProps>;
+
+    /**
+     * Whether the filtering state should be reset to initial when an item is selected
+     * (immediately before `onItemSelect` is invoked), or when the popover closes.
+     * The query will become the empty string and the first item will be made active.
+     * @default true
+     */
+    resetOnSelect?: boolean;
+
+    /** Controlled selected values. */
+    selectedItems?: T[];
+
+    /** Props to spread to `TagInput`. */
+    tagInputProps?: Partial<ITagInputProps>;
+
+    /** Custom renderer to transform an item into a string for tags. */
+    tagRenderer: (item: T) => string;
+}
+
+export interface IMultiSelectItemRendererProps<T> extends ISelectItemRendererProps<T> {
+    /**
+     * Whether the item is selected.
+     */
+    isSelected?: boolean;
+}
+
+export interface IMultiSelectState<T> {
+    activeItem?: T;
+    isOpen?: boolean;
+    query?: string;
+    selectedItems?: T[];
```

Comment:
i would argue for removing this from state entirely and supporting _only_ controlled usage. the other state fields are about internal interactions, but `selectedItems` is the big deal and the user will want to track it anyway. this'll make usage and testing much simpler and more reliable.

note how `Select<T>` doesn't even have a prop for its value, just `onItemSelect` event handler for user to receive updates and manage their own state.